plugin-localization/src/main/java/com.mapbox.mapboxsdk.plugins.localization/LocalizationPlugin.java

```diff
@@ -0,0 +1,193 @@
+   * @since 0.1.0
+   */
+  public LocalizationPlugin(@NonNull MapboxMap mapboxMap) {
+    this.mapboxMap = mapboxMap;
+    setMapTextLanguage(getDeviceLanguage());
+  }
+
+  /**
+   * Retrieve the language that the device is currently set to.
+   *
+   * @return The correct code according to ISO-639-1
+   */
+  public String getDeviceLanguage() {
+    if (Build.VERSION.SDK_INT >= 21) {
+      String languageTag = Locale.getDefault().toLanguageTag();
+      if (languageTag.contains("Hans")) {
+        return "zh-Hans";
+      } else if (languageTag.contains("Hant")) {
+        return "zh";
+      } else {
+        return Locale.getDefault().toLanguageTag().substring(0, 2);
+      }
+    } else {
+      if (Locale.getDefault().getLanguage().equalsIgnoreCase("IW")) {
+        return "he";
+      } else if (Locale.getDefault().getLanguage().equalsIgnoreCase("IN")) {
+        return "id";
+      } else if (Locale.getDefault().getLanguage().equalsIgnoreCase("JI")) {
+        return "yi";
+      } else if (Locale.getDefault().getLanguage().equalsIgnoreCase("JW")) {
+        return "jv";
+      } else if (Locale.getDefault().getLanguage().contains("zh-Hans")
+        || Locale.getDefault().getDisplayName().contains("简体")) {
+        return "zh-Hans";
+      } else if (Locale.getDefault().getLanguage().contains("zh-Hant")
+        || Locale.getDefault().getDisplayName().contains("繁體")) {
+        return "zh";
+      } else {
+        return Locale.getDefault().getLanguage();
+      }
+    }
+  }
+
+  /**
+   * Sets map text to the specified language
+   *
+   * @param languageToSetMapTo The language that to set the map text to.
+   * @since 0.1.0
+   */
+  public void setMapTextLanguage(String languageToSetMapTo) {
+    for (Source source : mapboxMap.getSources()) {
+      if (sourceIsFromMapbox(source)) {
+        for (Layer layer : mapboxMap.getLayers()) {
+          if (layerHasAdjustableTextField(layer)) {
+            if (((SymbolLayer) layer).getTextField().getValue().contains("name")) {
+              layer.setProperties(textField(String.format("{name_%s}", languageToSetMapTo)));
+            }
+            if (((SymbolLayer) layer).getTextField().getValue().contains("abbr")
+              && !getDeviceLanguage().equals("en")) {
+              layer.setProperties(textField(String.format("{name_%s}", languageToSetMapTo)));
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Checks whether the map's source is a source provided by Mapbox, rather than a custom source.
+   *
+   * @param singleSource an individual source object from the map
+   * @return true if the source is from a Mapbox vector source, false if it's a
+   * custom Studio data set or from elsewhere.
+   */
+  private boolean sourceIsFromMapbox(Source singleSource) {
+    return singleSource instanceof VectorSource
+      && ((VectorSource) singleSource).getUrl().contains("mapbox://mapbox.mapbox");
```

Comment:
Many Mapbox-designed styles – indeed, many styles designed in Studio – use a single composite source whose URL is `mapbox://mapbox.mapbox-terrain-v2,mapbox.mapbox-streets-v7`. By only checking if the URL begins with `mapbox://mapbox.mapbox`, this code may inadvertently snag sources that are just based on `mapbox.satellite` or `mapbox.terrain`.

[The iOS/macOS map SDK’s implementation](https://github.com/mapbox/mapbox-gl-native/blob/0eb4b5cc0ec494f4b8de6933d2eddf97b12a77f7/platform/darwin/src/MGLVectorSource.mm#L102-L109) considers the source to be the Mapbox Streets source – the only one this plugin can reliably localize – if both of the following conditions hold:

1. The URL scheme is `mapbox`; that is, the URL begins with `mapbox://`.
1. The URL’s hostname is a comma-delimited list containing either `mapbox.mapbox-streets-v7` or `mapbox.mapbox-streets-v6`. (The latter doesn’t matter so much at this point.)

Reply:
Thanks @1ec5 . Will adjust based on your comments. 🚀

Reply:
Although I don't create String arrays like the iOS/macOS map SDK’s implementation, I believe that the way I've handled scheme/hostname check will work just fine: 

https://github.com/mapbox/mapbox-plugins-android/blob/08d5ef3013c52e8e8d4feb57811eeda6b05a4458/plugin-localization/src/main/java/com.mapbox.mapboxsdk.plugins.localization/LocalizationPlugin.java#L109-L111

Reply:
Yeah, I think it’s robust enough. It’s unlikely that someone would create an account named `likemapbox` and upload a dataset named `mapbox-streets-v71` to it. It’s also going to be awhile before the cartography team gets a chance to release Mapbox Streets source v60.